### ISSUE TITLE
Option to `cargo run` to make it set the project dir as the current working directory

### ISSUE THREAD

**Issue by ilyagr** [rust-lang/cargo#13348]
### Problem

I set up a cargo alias in `.cargo/config.toml` as follows:

```
[alias]
gen-cli-reference = "run -- generate-cli-reference docs/cli-reference.md"
```

The goal is so that `cargo gen-cli-reference` would be a nice cross-platform way a developer can auto-generate some docs.

This only works from the root directory of the project, because `cargo run` runs the binary in the current working directory.

### Proposed Solution

I would like an option `cargo run --in-project-dir` that runs the binary from the root of the project, which is `$CARGO_MANIFEST_DIR`, I think. Then, I could define

```
[alias]
gen-cli-reference = "run --in-project-dir -- generate-cli-reference docs/cli-reference.md"
```

and `cargo gen-cli-reference` would work from anywhere.

### Notes

**Workaround:** An alternative would be to write an [`xtask`](https://github.com/matklad/cargo-xtask) that reads `$CARGO_MANIFEST_DIR`. The `xtask` would then execute `cargo run`.

**Comment by epage** [comment]
For myself, this feels like a fairly specialized flag for run.  I wonder if another approach is to design this into a more advanced "alias" system that I would one day love to have.  Unfortunately, that is less firmly defined and very low on most people's priority lists.

**Comment by weihanglo** [comment]
Another issue in this proposal is the semantics of `--in-project-dir`. Does it refer to the workspace root or thee package root directory? I believe both of them are reasonable choices, but you cannot satisfy both usages with only one boolean flag.

And agree with epage this is fairly specialized and a bit low in priority. `xtask` is indeed a good alternative, as you've mentioned. Cargo itself also [makes use of it](https://github.com/rust-lang/cargo/blob/f4bafa7f9180224e970185eec082dcb828542d02/crates/xtask-build-man/src/main.rs#L80-L84) to determine package/workspace root. If you'd like to get the directory rustc is invoked from, see the unstable [`CARGO_RUSTC_CURRENT_DIR`](https://github.com/rust-lang/cargo/issues/3946#issuecomment-1832514876) env var.

Some other issues might be related:

* https://github.com/rust-lang/cargo/issues/6575
* https://github.com/rust-lang/cargo/issues/10789
* https://github.com/rust-lang/cargo/issues/12477


Since there is an alternative and nothing blocks people from using that, I'll label this as `P-low` as this moment. Please let us know if there is more info you'd like to add :)

**Comment by ilyagr** [comment]
I think this is fair, thanks for the example and the suggestions!

Another thing I missed is that `run -- generate-cli-reference > docs/cli-reference.md` couldn't be an alias. OTOH, I'm not sure whether that should be in-scope.

**Comment by weihanglo** [comment]
When you run `cargo run -- generate-cli-reference > docs/cli-reference.md` in your shell, `>` is parsed and interpreted as redirections by the shell. However, when it is an alias, there is no shell involved in parsing it in Cargo, so a list of args `["generate-cli-reference", ">", "docs/cli-reference.md"]` will be passed to the running executable I believe. This is discussed in <https://github.com/rust-lang/cargo/issues/6575>.

**Comment by ilyagr** [comment]
Yes, it'd need a different syntax, or some other solution. Shelling out would defeat the purpose of portability to Windows.

**Comment by tisonkun** [comment]
I'm running into this and find a proper workaround:

```toml
[env]
CARGO_WORKSPACE_DIR = { value = "", relative = true }

[alias]
o = "run --manifest-path $CARGO_WORKSPACE_DIR/dev/Cargo.toml --"
```

However, it seems alias string doesn't resolve env var and thus this fails to work. But perhaps the direction is worth to dig more.

**Comment by weihanglo** [comment]
There is no variable substitution in `[alias]` yet. We'll need to define a proper cross-platform escape for that. See <https://github.com/rust-lang/cargo/issues/13348#issuecomment-1912350511> for relevant issues and discussions.

**Comment by epage** [comment]
We talked about this briefly in the Cargo team meeting.    We felt that this likely needs an RFC that also includes #12477.  Something that would need to be figured out is how table merging works.

We'd need to explore what can be used for current dir
- e.g. config parent dir, workspace, package, etc
  - one word of caution on workspace/package is that we don't have those normally evaluated at the time of aliases and we likely only want to do this look up if used
- enumerated values, variable substitution, etc

Note that another workaround is to use `run -C` so long as we had variable substitution on the command line (#10098)

**Comment by epage** [comment]
#10789 is for more variable substitution in config.
